### PR TITLE
Potential Optimisation, depending on target .net framework.

### PR DESCRIPTION
**Pull request by AdamSpeight2008**
Conditionally define a different implementation of the extension function `TransitiveClosure(Of T)`, depending on the target .net framework.

This allows to the use of some pooled objects.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Utilities/FunctionExtensions.cs [hunk 1]
```diff
@@ -12,9 +13,35 @@ namespace Microsoft.CodeAnalysis.CSharp
 {
     internal static partial class FunctionExtensions
     {
+#if (NETCOREAPP)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Utilities/FunctionExtensions.cs:16]: i don't understand why we need both of these functions.

File: src/Dependencies/PooledObjects/PooledHashSet.cs [hunk 2]
```diff
@@ -43,4 +43,38 @@ public static PooledHashSet<T> GetInstance()
             return instance;
         }
     }
+
+    internal sealed partial class PooledStack<T> : Stack<T>
```
Comment (CyrusNajmabadi) [inline src/Dependencies/PooledObjects/PooledHashSet.cs:47]: shouldn't need this.  Can just use ArrayBuilder.

File: src/Compilers/VisualBasic/Portable/Utilities/FunctionExtensions.vb [hunk 3]
```diff
@@ -14,6 +14,26 @@ Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
...
+                    If closure.Add(newItem) Then
+                        stack.Push(newItem)
+                    End If
+                Next
+            End While
+            Dim result = closure.ToHashSet()
+            closure.Free()
+            Return result
+        End Function
+#Else
```
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/Utilities/FunctionExtensions.vb:36]: nothing about this seems vb specific.  why not have a single shared impl defined in common layer?

File: src/Compilers/VisualBasic/Portable/Utilities/FunctionExtensions.vb [hunk 4]
```diff
@@ -14,6 +14,26 @@ Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
 'TODO - This is copied from C# and should be moved to common assemble.
 Namespace Microsoft.CodeAnalysis.VisualBasic
     Friend Module FunctionExtensions
+
+#If NETCOREAPP Then
+        <Extension()>
+        Public Function TransitiveClosure(Of T)(relation As Func(Of T, IEnumerable(Of T)), item As T) As HashSet(Of T)
+            Dim closure = PooledObjects.PooledHashSet(Of T).GetInstance
+            Dim stack as New Stack(Of T)()
```
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/Utilities/FunctionExtensions.vb:22]: why is htis not using pooledstack?

File: src/Compilers/VisualBasic/Portable/Utilities/FunctionExtensions.vb (hunk 3 above)
Comment (AdamSpeight2008) [reply src/Compilers/VisualBasic/Portable/Utilities/FunctionExtensions.vb:36]: @CyrusNajmabadi I think a better question would be. Why is this method not called by anything else in VB, where as C#'s is?

File: src/Compilers/CSharp/Portable/Utilities/FunctionExtensions.cs (hunk 1 above)
Comment (AdamSpeight2008) [reply src/Compilers/CSharp/Portable/Utilities/FunctionExtensions.cs:16]: Extension method `ToHash<T>` is not available in `.netstandard2.0`.

File: src/Dependencies/PooledObjects/PooledHashSet.cs (hunk 2 above)
Comment (AdamSpeight2008) [reply src/Dependencies/PooledObjects/PooledHashSet.cs:47]: Not if we are need a pooled collection of `Stack<T>`.

File: src/Compilers/VisualBasic/Portable/Utilities/FunctionExtensions.vb (hunk 3 above)
Comment (AdamSpeight2008) [reply src/Compilers/VisualBasic/Portable/Utilities/FunctionExtensions.vb:36]: re #45330

File: src/Compilers/VisualBasic/Portable/Utilities/FunctionExtensions.vb (hunk 4 above)
Comment (AdamSpeight2008) [reply src/Compilers/VisualBasic/Portable/Utilities/FunctionExtensions.vb:22]: Didn't get round to updating it yet. Plus no point if going shared common implementation.

**Pull request closed without merging**